### PR TITLE
add new shared module for authn to registry.redhat.io

### PR DESCRIPTION
**Pull request by smccarthy-ie**
See https://issues.redhat.com/browse/FUSEDOC-4019

### CODE REVIEW

**Comment by smccarthy-ie** [comment]
@bhardesty @ncbaratta When you get a chance, can you a review this new shared module? Thanks

File: modules/proc-configuring-container-registry-authn.adoc [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+// Attributes used:
+//
+// {ProductName}
+// {OCPVersion}
+
+
+[id="configuring-container-registry-authn_{context}"]
+= Configuring authentication with registry.redhat.io
+
+You must configure authentication with `registry.redhat.io` before you can deploy {ProductName} container images on OpenShift.
```
Comment (ncbaratta) [inline modules/proc-configuring-container-registry-authn.adoc]: missing the abstract tag

Comment (ncbaratta) [inline modules/proc-configuring-container-registry-authn.adoc]: I'd change this from 'you must' to a more abstracty type sentence.  'Configure authentication with ....'

File: modules/proc-configuring-container-registry-authn.adoc [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+
+.Prerequisites
+* You must have cluster administrator access to an OpenShift Container Platform cluster.
```
Comment (ncbaratta) [inline modules/proc-configuring-container-registry-authn.adoc]: Depends I guess on the style you use in your guides, but we avoid ' you' in front of these.

```suggestion
* Have cluster administrator access to an OpenShift Container Platform cluster.  
```

File: modules/proc-configuring-container-registry-authn.adoc [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+
+
+[id="configuring-container-registry-authn_{context}"]
+= Configuring authentication with registry.redhat.io
+
+You must configure authentication with `registry.redhat.io` before you can deploy {ProductName} container images on OpenShift.
+
+.Prerequisites
+* You must have cluster administrator access to an OpenShift Container Platform cluster.  
+* You must have the OpenShift `oc` client tool installed. For more details, see the link:https://docs.openshift.com/container-platform/{OCPVersion}/cli_reference/openshift_cli/getting-started-cli.html[OpenShift CLI documentation].
```
Comment (ncbaratta) [inline modules/proc-configuring-container-registry-authn.adoc]:
```suggestion
* Have the OpenShift `oc` client tool installed. For more details, see the link:https://docs.openshift.com/container-platform/{OCPVersion}/cli_reference/openshift_cli/getting-started-cli.html[OpenShift CLI documentation]. 
```

File: modules/proc-configuring-container-registry-authn.adoc [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+
+You must configure authentication with `registry.redhat.io` before you can deploy {ProductName} container images on OpenShift.
+
+.Prerequisites
+* You must have cluster administrator access to an OpenShift Container Platform cluster.  
+* You must have the OpenShift `oc` client tool installed. For more details, see the link:https://docs.openshift.com/container-platform/{OCPVersion}/cli_reference/openshift_cli/getting-started-cli.html[OpenShift CLI documentation]. 
+
+.Procedure
+
+. Log into your OpenShift cluster as administrator, for example:
```
Comment (ncbaratta) [inline modules/proc-configuring-container-registry-authn.adoc]:
```suggestion
. Log into your OpenShift cluster as administrator

File: modules/proc-configuring-container-registry-authn.adoc [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
++
+----
+$ oc login --user system:admin --token=my-token --server=https://my-cluster.example.com:6443
+----
+
+. Open the project in which you want to deploy {ProductName}, for example:
```
Comment (ncbaratta) [inline modules/proc-configuring-container-registry-authn.adoc]:
```suggestion
. Open the project in which you want to deploy {ProductName}
```

File: modules/proc-configuring-container-registry-authn.adoc [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
++
+You should see the following output:
++
+----
+secret/PULL_SECRET_NAME created
+----
++
+[IMPORTANT]
+====
+You must create this pull secret in every OpenShift project namespace that will authenticate to `registry.redhat.io`.
```
Comment (ncbaratta) [inline modules/proc-configuring-container-registry-authn.adoc]:
```suggestion
You must create this `docker-registry` secret in every OpenShift project namespace that will authenticate to `registry.redhat.io`.
```

File: modules/proc-configuring-container-registry-authn.adoc [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+. Link the secret to the `builder` service account in the namespace in which you will deploy the images: 
++
+----
+$ oc secrets link builder PULL_SECRET_NAME
+----
++
+NOTE: If you do not want to use your Red Hat username and password to create the pull secret, you can create an authentication token using a registry service account.
+
+.Additional resources
+* https://access.redhat.com/RegistryAuthentication[Red Hat container image authentication]
```
Comment (ncbaratta) [inline modules/proc-configuring-container-registry-authn.adoc:82]: This should be a sentence - "for more information about .... see .... "

File: modules/proc-configuring-container-registry-authn.adoc [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+* https://access.redhat.com/terms-based-registry/[Red Hat registry service accounts]
```
Comment (ncbaratta) [inline modules/proc-configuring-container-registry-authn.adoc:83]: This should be a sentence - "for more information about .... see .... "

**Review by ncbaratta — CHANGES_REQUESTED** [review]

**Comment by smccarthy-ie** [comment]
Thanks @ncbaratta for the excellent review feedback. Very helpful, all done now. Let me know if you spot anything else before merging.

File: modules/proc-configuring-container-registry-authn.adoc [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+// @smccarthy-ie - Getting Started with Camel Kafka Connector
+//
+// Attributes used:
+//
+// {ProductName}
+// {OCPVersion}
+
+
+[id="configuring-container-registry-authn_{context}"]
+= Authenticating with Red Hat for container images
```
Comment (oraNod) [inline modules/proc-configuring-container-registry-authn.adoc]: Authenticating with the Red Hat Container Registry maybe? or Ecosystem Catalog

File: modules/proc-configuring-container-registry-authn.adoc [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+
+.Prerequisites
+* Cluster administrator access to an OpenShift Container Platform cluster.  
+* OpenShift `oc` client tool is installed. For more details, see the link:https://docs.openshift.com/container-platform/{OCPVersion}/cli_reference/openshift_cli/getting-started-cli.html[OpenShift CLI documentation]. 
+
+.Procedure
+
+. Log into your OpenShift cluster as administrator:
++
+----
```
Comment (oraNod) [inline modules/proc-configuring-container-registry-authn.adoc:26]: syntax highlighting? [source,yaml,options="nowrap"]

File: modules/proc-configuring-container-registry-authn.adoc [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+secret/PULL_SECRET_NAME created
+----
++
+[IMPORTANT]
+====
+You must create this `docker-registry` secret in every OpenShift project namespace that will authenticate to `registry.redhat.io`.
+====
++
```
Comment (oraNod) [inline modules/proc-configuring-container-registry-authn.adoc]: extra + ?

**Comment by oraNod** [comment]
@smccarthy-ie and @ncbaratta hey! I've been thinking about this PR and would like to share some thoughts and maybe ask some q's. I don't know your plans for this module so don't mind me too much but my first thought is about overlap with instructions for the container image itself.

I don't know if this is the correct one but, as an example: https://catalog.redhat.com/software/containers/integration-tech-preview/camel-k-rhel8-operator/5e9d5f795a134668769d542d?container-tabs=gti&gti-tabs=registry-tokens 

Maybe it would be easier to maintain a url in the doc set than a set of instructions? I've now realized that the DG 7.3 docs need to be updated because they've become out of sync with the catalog. 

Another thing I wonder about is how well these instructions match production. Probably fine for local dev but I'd imagine Ansible or some other automation tooling would be used. Not suggesting anything for this PR. But is this ever something you'd look at? 

Cheers.

File: modules/proc-configuring-container-registry-authn.adoc (hunk 9 above)
Comment (smccarthy-ie) [reply modules/proc-configuring-container-registry-authn.adoc]: Thanks Don. Yes, this is an interesting one. There was an initiative to clarify registry naming and the latest recommendation is very generic, e.g., "pulling images from Red Hat" (see https://docs.google.com/document/d/1kA633C0XwvWXpsZyuDxwrWLdE85HbR9SfVqq2EuNH-s/edit#). I've changed it back to authenticating with `registry.redhat.io`, which does exactly what it says on the tin :)

Comment (oraNod) [reply modules/proc-configuring-container-registry-authn.adoc]: TIL. thanks, Stephen!

File: modules/proc-configuring-container-registry-authn.adoc (hunk 11 above)
Comment (smccarthy-ie) [reply modules/proc-configuring-container-registry-authn.adoc]: Good spot, thanks :)

**Comment by smccarthy-ie** [comment]
Hey @oraNod + @ncbaratta, 

Yes, good questions. There definitely are URLs that you can point to that have some of the steps required to authenticate to `registry.redhat.io`. For RH Integration products like Camel Kafka Connector, Fuse Online, 3scale, and AMQ Streams, there are some specific additional steps required to link the pull secret to your service account and to the builder service account for building images. 

And if you don't perform these steps, you cannot deploy these products. I wasted a lot of time trying to figure this out as a new Camel Kafka Connector user, and then noticed the same applied to several other RH Integration products (see https://issues.redhat.com/browse/FUSEDOC-4019). So I think this module is a useful addition for the Middleware Integration products at least. I don't know if it would apply to many other Middleware products.

**Comment by smccarthy-ie** [comment]
@oraNod I need to think about your second question on automation. These instructions are targeted at the getting started use case. I guess they show what's required if you have the smarts to automate it :)

**Comment by oraNod** [comment]
@smccarthy-ie Thanks for explaining! Definitely makes sense to have this as a shared module. Data Grid is supported on OpenShift via the operator so the steps to pull images don't apply but you do for 7.3, but that's going into ELS. So I wouldn't bother single-sourcing for that version. If pulling DG images becomes supported down the line I'll come back to this though. 

I guess my automation comment is something I find intriguing more than anything. Docs usually need to address the getting started use case and we can rely on a degree of competence for the power users out there writing Ansible playbooks. At the end of the day those are just instructions in yaml though.

**Comment by smccarthy-ie** [comment]
@ncbaratta Please let me know if there's anything else we need to do to get this one merged? I'd like to share this module with the the RH Integration doc writers and close https://issues.redhat.com/browse/FUSEDOC-4019 :) Thanks

**Comment by smccarthy-ie** [comment]
@fbolton This is the PR for https://issues.redhat.com/browse/FUSEDOC-4019. Please let me know if you have any review feedback or are happy to merge? Thanks.

**Comment by fbolton** [comment]
Looks like all of the review feedback on this MR has been addressed already. I will go ahead and merge it now.

**Comment by ncbaratta** [comment]
> @ncbaratta Please let me know if there's anything else we need to do to get this one merged? I'd like to share this module with the the RH Integration doc writers and close https://issues.redhat.com/browse/FUSEDOC-4019 :) Thanks

Sorry!! Glad it got merged.

**Pull request merged**